### PR TITLE
feat: Initialization process improvements

### PR DESCRIPTION
**Pull request by nicolantean**
Improvements in the dependencies injection process.

### CODE REVIEW

File: lib/core/di/di_utils_module.dart [hunk 1]
```diff
@@ -11,10 +12,21 @@ class UtilsDiModule {
...
+    await locator.allReady();
+  }
 }
 
 extension _GetItUseCaseDiModuleExtensions on GetIt {
   void _setupModule() {
     registerSingleton(AppRouter());
   }
+
+  void _setupProviders() {
```
Comment (f7deleon) [inline lib/core/di/di_utils_module.dart:27]: expression?

File: lib/core/di/di_utils_module.dart [hunk 2]
```diff
@@ -11,10 +12,21 @@ class UtilsDiModule {
   void setupModule(GetIt locator) {
     locator._setupModule();
   }
+
+  Future<void> setupProviders(GetIt locator) async {
+    locator._setupProviders();
+    await locator.allReady();
```
Comment (f7deleon) [inline lib/core/di/di_utils_module.dart:18]: its ok to call `allReady` two times?

**Review by f7deleon — APPROVED** [review]

File: lib/core/di/di_utils_module.dart (hunk 2 above)
Comment (mirland) [reply lib/core/di/di_utils_module.dart:18]: Yes, I think it's ok. 
the providers are the only thing that they are async, so you need the providers to initialize the other things.

The order is:
1. providers 
2. other things

File: lib/core/di/di_utils_module.dart [hunk 3]
```diff
@@ -11,10 +12,21 @@ class UtilsDiModule {
...
+  }
 }
 
 extension _GetItUseCaseDiModuleExtensions on GetIt {
```
Comment (mirland) [inline lib/core/di/di_utils_module.dart:22]: rename it

File: lib/core/di/di_utils_module.dart [hunk 4]
```diff
@@ -11,10 +12,21 @@ class UtilsDiModule {
   void setupModule(GetIt locator) {
     locator._setupModule();
   }
+
+  Future<void> setupProviders(GetIt locator) async {
```
Comment (mirland) [inline lib/core/di/di_utils_module.dart:16]: what do you think about adding it in a new file?

File: lib/core/di/di_provider.dart [hunk 5]
```diff
@@ -3,21 +3,15 @@ import 'package:flutter_template/core/di/di_utils_module.dart';
...
-    final instance = GetIt.instance;
-    if (!_initialized) {
-      RepositoryDiModule().setupModule(instance);
-      UtilsDiModule().setupModule(instance);
-      _initialized = true;
-    }
-
-    return instance;
+  static Future<void> init() async {
+    await UtilsDiModule().setupProviders(_instance);
```
Comment (mirland) [inline lib/core/di/di_provider.dart:9]: add a comment that it should be done first

**Review by mirland — APPROVED** [review]

File: lib/core/di/di_utils_module.dart (hunk 2 above)
Comment (mirland) [reply lib/core/di/di_utils_module.dart:18]: > but if they are async and the other are async shouldn't that be fixed with just one allReady?

The issue is that for example you cannot initialize the localsource if you didn't initialize the DB

So, as the DB initialization is async you have to call

await dbInit
await localSource

in this example is like>

await providers
await other things (repositories, sources, utiles, etc)

File: lib/core/di/di_provider.dart [hunk 6]
```diff
@@ -1,23 +1,19 @@
...
-    final instance = GetIt.instance;
-    if (!_initialized) {
-      RepositoryDiModule().setupModule(instance);
-      UtilsDiModule().setupModule(instance);
-      _initialized = true;
-    }
-
-    return instance;
+  static Future<void> init() async {
+    // Setup app providers should be done first
```
Comment (mirland) [inline lib/core/di/di_provider.dart:10]:
```suggestion
    // Setup app providers have to be done first
```

File: lib/core/di/di_utils_module.dart [hunk 7]
```diff
@@ -8,12 +8,13 @@ class UtilsDiModule {
 
   factory UtilsDiModule() => _instance;
 
-  void setupModule(GetIt locator) {
+  Future<void> setupModule(GetIt locator) async {
     locator._setupModule();
+    await locator.allReady();
```
Comment (mirland) [inline lib/core/di/di_utils_module.dart:13]: not needed

File: lib/core/di/di_utils_module.dart [hunk 8]
```diff
@@ -8,12 +8,13 @@ class UtilsDiModule {
 
   factory UtilsDiModule() => _instance;
 
-  void setupModule(GetIt locator) {
+  Future<void> setupModule(GetIt locator) async {
```
Comment (mirland) [inline lib/core/di/di_utils_module.dart:11]: remove async

**Review by mirland — APPROVED** [review]

**Review by mirland — APPROVED** [review]

**Pull request merged**